nova_vm/src/lib.rs

```diff
@@ -0,0 +1,576 @@
+            Value::Symbol(_) | Value::BigInt(_) => todo!("type error"),
+            Value::Undefined => f64::NAN,
+            Value::Null | Value::Boolean(false) => 0.,
+            Value::Boolean(true) => 1.,
+            Value::String(_) => todo!("parse number from string"),
+            _ => todo!("should assert as object and do other steps"),
+        })
+    }
+}
+
+type JsResult<T> = std::result::Result<T, Value>;
+
+impl Debug for Value {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Null => write!(f, "Null"),
+            Self::Undefined => write!(f, "Undefined"),
+            Self::Boolean(arg0) => f.debug_tuple("Boolean").field(arg0).finish(),
+            Self::Number(arg0) => f.debug_tuple("Number").field(arg0).finish(),
+            Self::BigInt(arg0) => f.debug_tuple("BigInt").field(arg0).finish(),
+            Self::String(arg0) => f.debug_tuple("String").field(&arg0.data.as_str()).finish(),
+            Self::Object(arg0) => f.debug_tuple("JsObject").field(&arg0).finish(),
+            Self::Symbol(arg0) => f
+                .debug_tuple("Symbol")
+                .field(&arg0.descriptor.as_ref().map(|wtf8| wtf8.data.as_str()))
+                .finish(),
+        }
+    }
+}
+
+pub trait JsObject: Trace + Debug {}
+
+#[derive(Clone)]
+pub struct JsString {
+    data: Wtf8Buf,
+}
+
+impl Finalize for JsString {
+    fn finalize(&self) {}
+}
+unsafe impl Trace for JsString {
+    unsafe_empty_trace!();
+}
+
+#[derive(Trace, Finalize)]
+pub struct JsSymbol {
+    descriptor: Option<Gc<JsString>>,
+}
+
+#[repr(u32)]
+pub enum Instruction {
+    LoadInteger,
+    LoadBoolean,
+    LoadNull,
+    LoadString,
+
+    // [out] [in]
+    CopyValue,
+
+    // [out] [left] [right]
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Mod,
+    StrictEquality,
+    Equality,
+    StrictInequality,
+    Inequality,
+
+    /// `[jump_rel]`
+    ///
+    /// Jumps to the relative instruction.
+    Jump,
+
+    /// `[addr] [jump_rel]`
+    ///
+    /// If `addr` is a true when converted to a boolean, then the instruction
+    /// will skip `jump_rel` instructions forward.
+    Test,
+
+    /// `[addr] [jump_rel]`
+    /// If `addr` is false when converted to a boolean, then the instruction
+    /// will skip `jump_rel` instructions forward.
+    TestNot,
+}
+
+impl Into<u32> for Instruction {
+    fn into(self) -> u32 {
+        self as u32
+    }
+}
+
+pub struct VM<'a> {
+    pub source: &'a str,
+    pub instructions: Vec<u32>,
+    /// Program counter.
+    pub pc: u32,
+    pub strings: Vec<Gc<JsString>>,
+}
+
+#[derive(Debug, Default)]
+pub struct Env<'a> {
+    pub map: std::collections::HashMap<&'a str, u32>,
+    pub parent: Option<Box<Env<'a>>>,
+}
+
+impl<'a> VM<'a> {
+    pub fn interpret(&self) -> JsResult<()> {
+        let mut memory = Vec::<Value>::with_capacity(self.pc as usize);
+
+        for _ in 0..self.pc {
+            memory.push(Value::Undefined);
```

Comment:
Why is this needed, since each instruction can specify the memory address it operates on?